src/zenml/integrations/s3/artifact_stores/s3_artifact_store.py

```diff
@@ -467,10 +465,27 @@ def _remove_previous_file_versions(self, path: PathType) -> None:
             if isinstance(path, bytes):
                 path = path.decode()
             _, prefix = split_s3_path(path)
-            s3 = boto3.resource("s3")
-            bucket = s3.Bucket(self.config.bucket)
-            for version in bucket.object_versions.filter(Prefix=prefix):
+            for version in self._boto3_bucket.object_versions.filter(
+                Prefix=prefix
+            ):
                 if not version.is_latest:
                     version.delete()
 
         return
+
+    @property
+    def _boto3_bucket(self) -> Any:
+        """Get the boto3 bucket object.
+
+        Returns:
+            The boto3 bucket object.
+        """
+        key, secret, token, region = self.get_credentials()
+        s3 = boto3.resource(
+            "s3",
+            aws_access_key_id=key,
+            aws_secret_access_key=secret,
+            aws_session_token=token,
+            region_name=region,
+        )
+        return s3.Bucket(self.config.bucket)
```

Comment:
I had a hunch and asked ChatGPT about it. Here's what it said, but I couldn't verify it yet:

In your current implementation, every time _boto3_bucket() is called, it will create a new S3 connection via boto3.resource. This can lead to inefficiencies, especially in scenarios where the function is called frequently. To avoid creating a new connection on every call, you can optimize it by caching or reusing the S3 resource object.

One of the simplest ways to avoid creating new connections repeatedly is to store the boto3.resource object as an instance variable (in case you're working within a class) or as a global variable. This way, you only create the S3 resource once and reuse it across multiple calls.

```
    def __init__(self):
        self._s3_resource = None  # Cache the S3 resource object

    def _boto3_bucket(self) -> Any:
        """Get the boto3 bucket object.
        Returns:
            The boto3 bucket object.
        """
        if self._s3_resource is None:  # Create the resource only once
            key, secret, token, region = self.get_credentials()
            self._s3_resource = boto3.resource(
                "s3",
                aws_access_key_id=key,
                aws_secret_access_key=secret,
                aws_session_token=token,
                region_name=region,
            )
        return self._s3_resource.Bucket(self.config.bucket)
```


However, it gets a bit more complicated now, because the service connector credentials themselves can expire, so you'd need to refresh the resource object if that happens:

```
    def _boto3_bucket(self) -> Any:
        """Get the boto3 bucket object.
        Returns:
            The boto3 bucket object.
        """
        # Refresh the bucket resource also if the connector has expired
        if self._s3_resource and not self.connector_has_expired():
            return self._s3_resource.Bucket(self.config.bucket)

        key, secret, token, region = self.get_credentials()
        self._s3_resource = boto3.resource(
            "s3",
            aws_access_key_id=key,
            aws_secret_access_key=secret,
            aws_session_token=token,
            region_name=region,
        )
        return self._s3_resource.Bucket(self.config.bucket)
```